src/libs/opts/opts.c

```diff
@@ -0,0 +1,951 @@
+/**
+ * @file
+ *
+ * @brief
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ */
+
+#include <kdbopts.h>
+
+#include <stdlib.h>
+#include <string.h>
+
+#include <kdbease.h>
+#include <kdbhelper.h>
+#include <kdbmeta.h>
+
+#include <kdberrors.h>
+
+#ifdef _WIN32
+static const char SEP_ENV_VALUE = ';';
+#else
+static const char SEP_ENV_VALUE = ':';
+#endif
+
+static inline const char * keyGetMetaString (const Key * key, const char * meta)
+{
+	const Key * mk = keyGetMeta (key, meta);
+	const char * value = mk == NULL ? NULL : keyString (mk);
+	return value != NULL && value[0] == '\0' ? NULL : value;
+}
+
+static int addProcKey (KeySet * ks, const Key * key, Key * valueKey);
+static KeySet * parseEnvp (const char ** envp);
+
+static KeySet * parseArgs (KeySet * optionsSpec, int argc, const char ** argv, Key * errorKey);
+static void setOption (Key * option, const char * value, bool repeated);
+
+static Key * splitEnvValue (const Key * envKey);
+
+static KeySet * ksMetaGetSingleOrArray (Key * key, const char * metaName);
+
+static char * generateHelpMessage (KeySet * optionsSpec, const char * progname);
+
+/**
+ * This functions parses a specification of program options, together with a list of arguments
+ * and environment variables to extract the option values.
+ *
+ * The options have to be defined in the metadata of keys in the spec namespace. If an option value
+ * is found for any of the given keys, a new key with the same path but inside the proc namespace
+ * will be inserted into @p ks. This enables a cascading lookup to find these values.
+ *
+ * If @p argv contains "-h" or "--help" @p ks will not be changed, instead the value of @p errorKey
+ * will be set to a help message describing available options and 1 will be returned. The program
+ * name used in this message is taken from `argv[0]`. If it contains a '/' only the part after the
+ * last '/' will be used. The help message will contain description for each option. This description
+ * is taken from `opt/help`, or if `opt/help` is not specified from `description`. If neither `opt/help`
+ * nor `description` are given, or `opt/nohelp` is set to "1", the option will not appear in the help
```

Comment:
> I think the option should be listed, even if neither opt/help nor description is given.

You are right, `opt/nohelp` should be good enough.

> What is the usecase for opt/nohelp? It is not described above?

 I will add documentation. clang for example has some hidden options, that are only mentioned when `--help-hidden` is used. It could also be used for deprecated options only supported for legacy reasons (i.e. to not break old scripts).